Suppose we want to predict whether it will be Sunny or Rainy tomorrow. We don't know the actual weather (hidden state) but can observe whether a person carries an umbrella or not (observed event). The states and observations can be defined as:

Hidden States: Sunny, Rainy
Observations: Umbrella, No Umbrella
We need a few parameters for our HMM:

Initial State Probabilities: Probability of the initial state (Sunny or Rainy).
Transition Probabilities: Probability of transitioning from one state to another (e.g., Sunny to Rainy, or Rainy to Sunny).
Emission Probabilities: Probability of observing an event given a state (e.g., observing 'Umbrella' given 'Rainy').
For simplicity, let's assume:

Initial State Probabilities: Sunny (0.5), Rainy (0.5)
Transition Probabilities:
Sunny to Sunny (0.7), Sunny to Rainy (0.3)
Rainy to Rainy (0.6), Rainy to Sunny (0.4)
Emission Probabilities:
Sunny: Umbrella (0.1), No Umbrella (0.9)
Rainy: Umbrella (0.8), No Umbrella (0.2)
Now, let's use Python to create this simple HMM. We'll use the hmmlearn library for this purpose. If hmmlearn is not installed, you can install it using pip (pip install hmmlearn). Let's write the Python code to model this HMM.

MultinomialHMM model from the hmmlearn library expects an additional parameter, n_trials, to be set for multinomial distributions. This parameter specifies the number of trials or the size of each sample in the observations.

In the context of our weather prediction example, since we are only observing one of two possible events (Umbrella or No Umbrella) at each time step, we can set n_trials to 1. Here's how you can modify the code to include this parameter:

In [15]:
from hmmlearn import hmm
import numpy as np

# Define the model parameters
states = ["Sunny", "Rainy"]
n_states = len(states)

observations = ["Umbrella", "No Umbrella"]
n_observations = len(observations)

# Start probability (initial state distribution)
start_probability = np.array([0.5, 0.5])

# Transition probability matrix (A)
transition_probability = np.array([
  [0.7, 0.3],  # Sunny to Sunny, Sunny to Rainy
  [0.4, 0.6],  # Rainy to Sunny, Rainy to Rainy
])

# Emission probability matrix (B)
emission_probability = np.array([
  [0.1, 0.9],  # Sunny: P(Umbrella|Sunny), P(No Umbrella|Sunny)
  [0.8, 0.2],  # Rainy: P(Umbrella|Rainy), P(No Umbrella|Rainy)
])

# Create the HMM
model = hmm.MultinomialHMM(n_components=n_states, n_iter=10,n_trials=1, tol=0.01)
model.startprob_ = start_probability
model.transmat_ = transition_probability
model.emissionprob_ = emission_probability

# Define an observation sequence with one-hot encoding
# Umbrella, No Umbrella, Umbrella
observed_sequence = np.array([[1, 0], [0, 1], [1, 0]])

# Predict the hidden states based on observed sequence
logprob, hidden_states = model.decode(observed_sequence, algorithm="viterbi")

# Translate hidden states to meaningful labels
predicted_weather = [states[state] for state in hidden_states]

print(predicted_weather, logprob)


MultinomialHMM has undergone major changes. The previous version was implementing a CategoricalHMM (a special case of MultinomialHMM). This new implementation follows the standard definition for a Multinomial distribution (e.g. as in https://en.wikipedia.org/wiki/Multinomial_distribution). See these issues for details:
https://github.com/hmmlearn/hmmlearn/issues/335
https://github.com/hmmlearn/hmmlearn/issues/340


['Rainy', 'Sunny', 'Rainy'] -3.365058335046282


The output of your Hidden Markov Model (HMM) consists of two parts: the sequence of predicted hidden states (['Rainy', 'Sunny', 'Rainy']) and the log probability of this sequence (-3.365058335046282).

Predicted Hidden States: ['Rainy', 'Sunny', 'Rainy']

This is the sequence of weather conditions (hidden states) that the model predicts as the most likely, given your observation sequence (Umbrella, No Umbrella, Umbrella).
The model predicts that on the first day it was likely rainy (since the person carried an umbrella), on the second day it was likely sunny (since the person did not carry an umbrella), and on the third day, it was likely rainy again (umbrella observed).
Log Probability: -3.365058335046282

This number represents the log likelihood of the predicted sequence of hidden states given the model and the observation sequence. It's a measure of how probable the model finds this sequence of states, given the data and the model parameters.
The value is in log scale. A higher value (closer to zero) indicates a higher probability.
The negative value is normal in log probability for sequences, especially when dealing with probabilities less than 1 (which is usually the case in probabilistic models like HMMs).
In summary, the HMM model analyzed the sequence of observations (carrying or not carrying an umbrella) and predicted the most likely corresponding weather conditions for each day, along with the likelihood of this sequence of conditions occurring. This kind of analysis can be useful in situations where you want to infer hidden states from observable events.

Related Problems
## Financial Market Analysis

Let's consider a scenario where we want to predict market trends. In this case, our hidden states could represent different market conditions, and our observations could be specific market indicators.

Hidden States (Market Conditions):
Bull Market (Rising market trends)
Bear Market (Falling market trends)
Observations (Market Indicators):
High Trading Volume
Low Trading Volume
We would then define the parameters of the HMM:

Initial State Probabilities: Probability of the initial market condition (Bull or Bear).
Transition Probabilities: Probability of transitioning from one market condition to another (e.g., Bull to Bear, or Bear to Bull).
Emission Probabilities: Probability of observing a market indicator given a market condition (e.g., observing 'High Trading Volume' given 'Bull Market').
For simplicity, let's assume:

Initial State Probabilities: Bull (0.5), Bear (0.5)
Transition Probabilities:
Bull to Bull (0.7), Bull to Bear (0.3)
Bear to Bull (0.4), Bear to Bear (0.6)
Emission Probabilities:
Bull: High Trading Volume (0.8), Low Trading Volume (0.2)
Bear: High Trading Volume (0.3), Low Trading Volume (0.7)
Using these parameters, you can set up an HMM to analyze sequences of trading volumes to predict whether the market is more likely in a Bull or Bear state. This model could help traders or financial analysts in making informed decisions based on the underlying trends indicated by trading volumes.

### Example: Credit Scoring and Risk Management
In this scenario, our hidden states could represent different levels of credit risk, and our observations could be specific customer financial behaviors or indicators.

Hidden States (Credit Risk Levels):
Low Risk (customer is likely to repay loans)
High Risk (customer is at risk of defaulting)
Observations (Financial Behaviors or Indicators):
Timely Payments (customer makes payments on time)
Missed Payments (customer misses payments)
We then define the HMM parameters:

Initial State Probabilities: Probability of the initial credit risk state (Low Risk or High Risk).
Transition Probabilities: Probability of transitioning from one credit risk level to another (e.g., Low Risk to High Risk, or High Risk to Low Risk).
Emission Probabilities: Probability of observing a financial behavior given a credit risk level (e.g., observing 'Timely Payments' given 'Low Risk').
For simplicity, let's assume:

Initial State Probabilities: Low Risk (0.6), High Risk (0.4)
Transition Probabilities:
Low Risk to Low Risk (0.8), Low Risk to High Risk (0.2)
High Risk to Low Risk (0.3), High Risk to High Risk (0.7)
Emission Probabilities:
Low Risk: Timely Payments (0.9), Missed Payments (0.1)
High Risk: Timely Payments (0.4), Missed Payments (0.6)
Using these parameters, you can set up an HMM to analyze sequences of customer payment behaviors to predict their current credit risk level. This model could be valuable for banks or credit agencies in assessing and managing the risk associated with lending.

### Example: Customer Behavior Analysis
In this scenario, our hidden states could represent different customer shopping patterns or segments, and our observations could be specific types of purchases or shopping behaviors.

Hidden States (Customer Shopping Patterns):
Regular Buyer (frequent, consistent purchases)
Occasional Buyer (infrequent, sporadic purchases)
Observations (Types of Purchases):
High-Value Purchase (customer makes expensive purchases)
Low-Value Purchase (customer makes inexpensive purchases)
We then define the HMM parameters:

Initial State Probabilities: Probability of the initial shopping pattern (Regular Buyer or Occasional Buyer).
Transition Probabilities: Probability of transitioning from one shopping pattern to another (e.g., Regular Buyer to Occasional Buyer, or vice versa).
Emission Probabilities: Probability of observing a type of purchase given a shopping pattern (e.g., observing 'High-Value Purchase' given 'Regular Buyer').
For simplicity, let's assume:

Initial State Probabilities: Regular Buyer (0.6), Occasional Buyer (0.4)
Transition Probabilities:
Regular Buyer to Regular Buyer (0.7), Regular Buyer to Occasional Buyer (0.3)
Occasional Buyer to Regular Buyer (0.4), Occasional Buyer to Occasional Buyer (0.6)
Emission Probabilities:
Regular Buyer: High-Value Purchase (0.6), Low-Value Purchase (0.4)
Occasional Buyer: High-Value Purchase (0.3), Low-Value Purchase (0.7)
Using these parameters, you can set up an HMM to analyze sequences of customer purchase types to predict their current shopping pattern. This model can be valuable for retailers or e-commerce platforms in understanding and predicting customer buying behaviors, leading to more effective marketing and sales strategies.

### Example: Credit Card Fraud Detection
In the context of credit card fraud detection, our hidden states could represent the types of card usage behavior, and our observations would be specific transactional activities.

Hidden States (Card Usage Behavior):
Legitimate Activity (normal, everyday transactions)
Fraudulent Activity (unusual or suspicious transactions)
Observations (Transactional Activities):
Small Transaction (common, small-scale purchases)
Large Transaction (high-value purchases)
Unusual Location Transaction (transactions in unusual locations)
Frequent Transactions (a high number of transactions in a short period)
We then define the HMM parameters:

Initial State Probabilities: Probability of the initial state of card usage (Legitimate or Fraudulent).
Transition Probabilities: Probability of transitioning from one type of card usage behavior to another (e.g., from Legitimate to Fraudulent, or vice versa).
Emission Probabilities: Probability of observing a specific transactional activity given a card usage behavior (e.g., observing 'Large Transaction' given 'Fraudulent Activity').
For simplicity, let's assume:

Initial State Probabilities: Legitimate Activity (0.95), Fraudulent Activity (0.05)
Transition Probabilities:
Legitimate to Legitimate (0.97), Legitimate to Fraudulent (0.03)
Fraudulent to Legitimate (0.05), Fraudulent to Fraudulent (0.95)
Emission Probabilities:
Legitimate Activity: Small Transaction (0.7), Large Transaction (0.2), Unusual Location (0.05), Frequent Transactions (0.05)
Fraudulent Activity: Small Transaction (0.1), Large Transaction (0.6), Unusual Location (0.2), Frequent Transactions (0.1)
Using these parameters, you can set up an HMM to analyze sequences of transactional activities to predict the current state of card usage. This model can be valuable for financial institutions in detecting and preventing credit card fraud.

### Example: Predictive Maintenance for Manufacturing Equipment
Scenario:
We aim to predict when manufacturing equipment might need maintenance or is at risk of failure.

Hidden States (Equipment Condition States):
Normal Operation (equipment is functioning normally)
Maintenance Needed (equipment is starting to exhibit signs of wear or minor issues)
Near Failure (equipment is at high risk of failure soon)
Observations (Machine Performance Indicators):
Normal Readings (normal sensor readings like temperature, vibration, etc.)
Slight Anomalies (slight deviations from normal in sensor readings)
Major Anomalies (significant deviations from normal in sensor readings)
We then define the HMM parameters:

Initial State Probabilities: Probability of the initial state of equipment condition.
Transition Probabilities: Probability of transitioning from one equipment condition to another (e.g., from Normal Operation to Maintenance Needed).
Emission Probabilities: Probability of observing specific machine performance indicators given an equipment condition.
For simplicity, let's assume:

Initial State Probabilities: Normal Operation (0.8), Maintenance Needed (0.15), Near Failure (0.05)
Transition Probabilities:
Normal to Normal (0.9), Normal to Maintenance (0.09), Normal to Near Failure (0.01)
Maintenance to Normal (0.5), Maintenance to Maintenance (0.4), Maintenance to Near Failure (0.1)
Near Failure to Normal (0.05), Near Failure to Maintenance (0.15), Near Failure to Near Failure (0.8)
Emission Probabilities:
Normal Operation: Normal Readings (0.85), Slight Anomalies (0.1), Major Anomalies (0.05)
Maintenance Needed: Normal Readings (0.6), Slight Anomalies (0.3), Major Anomalies (0.1)
Near Failure: Normal Readings (0.3), Slight Anomalies (0.4), Major Anomalies (0.3)
Using these parameters, an HMM can analyze sequences of machine performance indicators to predict the current state of equipment condition. This can be invaluable for manufacturing plants to proactively perform maintenance, thereby reducing downtime and costs.

### Example: Inventory Level Prediction in Supply Chain Management
Scenario:
The goal is to predict the state of inventory levels (e.g., overstocked, adequately stocked, understocked) based on observed sales and supply data.

Hidden States (Inventory Conditions):
Overstocked (too much inventory, risk of excess)
Adequately Stocked (optimal level of inventory)
Understocked (insufficient inventory, risk of stockouts)
Observations (Sales and Supply Indicators):
High Sales (sales volumes are higher than average)
Average Sales (sales volumes are around the average)
Low Sales (sales volumes are lower than average)
High Supply (high incoming inventory)
Low Supply (low incoming inventory)
These observations can be derived from sales data and supply chain logistics information.

We then define the HMM parameters:

Initial State Probabilities: Probability of the initial inventory condition.
Transition Probabilities: Probability of transitioning from one inventory condition to another (e.g., from Adequately Stocked to Understocked).
Emission Probabilities: Probability of observing specific sales and supply indicators given an inventory condition.
For simplicity, let's assume:

Initial State Probabilities: Overstocked (0.2), Adequately Stocked (0.5), Understocked (0.3)
Transition Probabilities:
Overstocked to Overstocked (0.6), Overstocked to Adequate (0.3), Overstocked to Understocked (0.1)
Adequately to Overstocked (0.2), Adequately to Adequate (0.6), Adequately to Understocked (0.2)
Understocked to Overstocked (0.1), Understocked to Adequate (0.3), Understocked to Understocked (0.6)
Emission Probabilities:
Overstocked: High Sales (0.2), Average Sales (0.3), Low Sales (0.5), High Supply (0.4), Low Supply (0.1)
Adequately Stocked: High Sales (0.4), Average Sales (0.4), Low Sales (0.2), High Supply (0.2), Low Supply (0.2)
Understocked: High Sales (0.6), Average Sales (0.2), Low Sales (0.2), High Supply (0.1), Low Supply (0.4)
Using these parameters, an HMM can analyze sequences of sales and supply indicators to predict the current state of inventory. This model can be invaluable for supply chain managers to make informed decisions about ordering and inventory management.

### Example: Customer Churn Prediction in a Subscription-Based Service
Scenario:
The aim is to predict the likelihood of customers churning (i.e., discontinuing their subscription) based on their interaction and usage patterns.

Hidden States (Customer Loyalty States):
Loyal Customer (unlikely to churn, consistently engaged)
At-Risk Customer (showing signs of reduced engagement, potential risk of churning)
Churning Customer (high likelihood of discontinuing the service)
Observations (Customer Interactions and Usage Patterns):
High Activity (frequent use of the service, high engagement)
Medium Activity (regular but less frequent use of the service)
Low Activity (infrequent use of the service)
Complaints (customer complaints or issues raised)
Positive Feedback (positive interactions or feedback from the customer)
These observations can be derived from usage data, customer support interactions, and feedback surveys.

We then define the HMM parameters:

Initial State Probabilities: Probability of the initial customer loyalty state.
Transition Probabilities: Probability of transitioning from one loyalty state to another (e.g., from Loyal Customer to At-Risk Customer).
Emission Probabilities: Probability of observing specific interaction and usage patterns given a customer loyalty state.
For simplicity, let's assume:

Initial State Probabilities: Loyal Customer (0.7), At-Risk Customer (0.2), Churning Customer (0.1)
Transition Probabilities:
Loyal to Loyal (0.8), Loyal to At-Risk (0.15), Loyal to Churning (0.05)
At-Risk to Loyal (0.2), At-Risk to At-Risk (0.6), At-Risk to Churning (0.2)
Churning to Loyal (0.05), Churning to At-Risk (0.25), Churning to Churning (0.7)
Emission Probabilities:
Loyal Customer: High Activity (0.6), Medium Activity (0.3), Low Activity (0.05), Complaints (0.03), Positive Feedback (0.02)
At-Risk Customer: High Activity (0.3), Medium Activity (0.4), Low Activity (0.25), Complaints (0.03), Positive Feedback (0.02)
Churning Customer: High Activity (0.1), Medium Activity (0.2), Low Activity (0.6), Complaints (0.05), Positive Feedback (0.05)
Using these parameters, an HMM can analyze sequences of customer interactions and usage patterns to predict the current loyalty state of a customer. This model can be invaluable for businesses in identifying at-risk customers early and taking proactive steps to reduce churn.

### Example: Market Trend Prediction for Algorithmic Trading
Scenario:
The aim is to predict market trends (bullish or bearish) to inform trading strategies in algorithmic trading.

Hidden States (Market Conditions):
Bull Market (market trend is upward)
Bear Market (market trend is downward)
Observations (Market Indicators):
Rising Prices (increasing stock prices or indices)
Falling Prices (decreasing stock prices or indices)
High Volume (high trading volumes, indicating strong market interest)
Low Volume (low trading volumes, possibly indicating uncertainty or disinterest)
These observations can be derived from historical market data, such as stock prices and trading volumes.

We then define the HMM parameters:

Initial State Probabilities: Probability of the initial market condition (Bull or Bear Market).
Transition Probabilities: Probability of transitioning from one market condition to another (e.g., from Bull to Bear Market).
Emission Probabilities: Probability of observing specific market indicators given a market condition.
For simplicity, let's assume:

Initial State Probabilities: Bull Market (0.5), Bear Market (0.5)
Transition Probabilities:
Bull to Bull (0.7), Bull to Bear (0.3)
Bear to Bull (0.4), Bear to Bear (0.6)
Emission Probabilities:
Bull Market: Rising Prices (0.6), Falling Prices (0.1), High Volume (0.2), Low Volume (0.1)
Bear Market: Rising Prices (0.2), Falling Prices (0.6), High Volume (0.1), Low Volume (0.1)
Using these parameters, an HMM can analyze sequences of market indicators to predict the current market condition. This model can be invaluable for algorithmic traders in adjusting their trading strategies based on predicted market trends.

### Example: Quality Control in Manufacturing
Scenario:
The goal is to monitor the quality of products during manufacturing and identify potential quality issues.

Hidden States (Quality States of the Product):
Good Quality (product meets all quality standards)
Minor Defects (product has minor quality issues but may still be acceptable)
Major Defects (product has significant quality issues and is unacceptable)
Observations (Quality Inspection Findings):
Pass (product passes a specific quality check)
Minor Issue (product shows a minor issue in a quality check)
Major Issue (product shows a major issue in a quality check)
These observations would typically come from various checkpoints in the manufacturing process where products are inspected for quality.

We then define the HMM parameters:

Initial State Probabilities: Probability of the initial quality state of a product batch.
Transition Probabilities: Probability of transitioning from one quality state to another (e.g., from Good Quality to Minor Defects).
Emission Probabilities: Probability of observing specific inspection findings given a quality state.
For simplicity, let's assume:

Initial State Probabilities: Good Quality (0.7), Minor Defects (0.2), Major Defects (0.1)
Transition Probabilities:
Good to Good (0.9), Good to Minor (0.09), Good to Major (0.01)
Minor to Good (0.1), Minor to Minor (0.8), Minor to Major (0.1)
Major to Good (0.05), Major to Minor (0.15), Major to Major (0.8)
Emission Probabilities:
Good Quality: Pass (0.95), Minor Issue (0.04), Major Issue (0.01)
Minor Defects: Pass (0.7), Minor Issue (0.25), Major Issue (0.05)
Major Defects: Pass (0.4), Minor Issue (0.4), Major Issue (0.2)
Using these parameters, an HMM can analyze sequences of inspection findings to predict the current quality state of the manufacturing process. This model can be invaluable for quality control managers to identify trends in product quality and address issues proactively.